text/3742-mutate-non-mutable-locals.md

```diff
@@ -0,0 +1,249 @@
+ - `cargo fix` performs more than just `mut` additions/removals, meaning other important warnings
+   may be suppressed as well
+
+## Addressing language inconsistency
+
+We could split the `mut` keyword into separate keywords which independently
+describe mutability and ownership, with the latter being something like `&uniq`
+or `&only`.  Such a change would likely be too dramatically breaking to be worth
+doing, however.
+
+# Prior art
+[prior-art]: #prior-art
+
+Reassignment or mutation of immutable bindings is considered
+an error by just about every programming language with the
+concept of immutability.  There are still some interesting
+cases, though.
+
+## C and C++
+
+Technically speaking, a conforming C compiler is allowed to ignore or emit warnings
+for reassignment-of-const, although MSVC, Clang, and GCC all treat it as an error.
+C has no "mutable borrow" operation, but the equivalent to mutably borrowing an
+immutable local compiles with warnings under the above three compilers:
+
+```c
+int const a = 12;
+int *pa = &a; // WARNING! initialization discards 'const' qualifier from pointer target type
+```
+
+C++ is more strict, and rejects the above example with an error.  In either language,
+it's UB to attempt to modify the value of `a`.
+
+C and C++ compilers can use `const` annotations on local variables
+for optimization purposes, but stricter reference semantics allow Rust
+to perform the same optimizations regardless of how locals are annotated.
+
+## Zig
+
+Zig has recently pushed in the _opposite_ direction,
+and now refuses to compile a program when variables are declared mutable but are never mutated:
+
+```zig
+var foo = 12; // ERROR! local variable is never mutated
+_ = foo;
+```
+
+The change was somewhat controversial, and many discussions about it can be found online.  Points raised in
+those discussions are very similar to the points raised here.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+ - How many lints should there be?  Reassignment and mutable borrowing
+   seem distinct enough to warrant separate lints.
+ - What name should the lint(s) have?
+    - `mut_non_mut`
+       - used in this document
+       - short for "mutate non-mutable"
+       - a bit cryptic
+    - `missing_mut`
+       - proposed in the IRLO thread
+       - suggests that `mut` can be added to fix, which isn't always true
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+There exist other candidates for "lintification," but they usually fall into one of
+two categories: either they don't cause enough friction to be worth changing, or successfully
+compiling the "error" case would require extra code generation.  For example:
+
+```rs
+if condition {
+    value // rejected due to missing semicolon: probably not worth changing.
+}
+```
+
+```rs
+fn maybe_thing(i: i32) -> Option<i32> {
+    i // rejected due to missing `Some`: Would require code generation to convert to a lint.
+}
+```
```

Comment:
This is also a type error.

I think you should move away from using the term "lintification". A lint is a well-formedness check on a program that's non-essential for soundness. `mut_non_mut` — even today — falls under this definition. This RFC only proposes that it be moved from forbid-by-default (which cannot be overridden), to deny-by-default (which can be overridden).

These "Future Possibilities" are an entirely different thing, proposing type-directed syntactic sugar. This RFC is controversial enough *without* these (im)possibilities. I suggest you remove them altogether.

Reply:
You're right, the examples were a distraction, so I removed them.  That said, they were based on [points raised in the IRLO thread](https://internals.rust-lang.org/t/lack-of-mut-in-bindings-as-a-deny-by-default-lint/15818/74), which might provide better context/explanation than I did.